src/bin/pg_waldump/pg_waldump.c

```diff
@@ -286,11 +286,26 @@ search_directory(const char *directory, const char *fname)
 
 			WalSegSz = longhdr->xlp_seg_size;
 
+			// if we skip errors, we don't need to check the segment size
 			if (!IsValidWalSegSize(WalSegSz))
+			{
+				if (!ignore_format_errors)
+				{
 				pg_fatal(ngettext("WAL segment size must be a power of two between 1 MB and 1 GB, but the WAL file \"%s\" header specifies %d byte",
 								  "WAL segment size must be a power of two between 1 MB and 1 GB, but the WAL file \"%s\" header specifies %d bytes",
 								  WalSegSz),
 						 fname, WalSegSz);
+				}
+				else
+				{
+					struct stat stat;
+					if(fstat(fd, &stat) != 0)
+						pg_fatal("could not stat file \"%s\"", fname);
+
+					/* Round up segment size to next power of 2 or 1MB */
+					WalSegSz = Max(next_pow2_int(stat.st_size), 1024 * 1024);
```

Comment:
I do not understand the logic here: we get file size and then round it on power of two.
Either segment file has valid size and there is no need to align it, either it can bar smaller than segment size just because only some part of this segment was written, but in this case alignment on power of two will not help.

Reply:
Fixed.

This shouldn't be an issue for neon, because we always write full segment, even for our special segment 0.